src/DataStructures/Tensor/Expressions/Evaluate.hpp

```diff
@@ -329,4 +368,58 @@ auto evaluate(const RhsTE& rhs_tensorexpression) {
                                 ~rhs_tensorexpression);
   return lhs_tensor;
 }
+
+/*!
+ * \ingroup TensorExpressionsGroup
+ * \brief Update a LHS tensor to the result of a RHS tensor expression that
+ * contains it
+ *
+ * \details See documentation for `evaluate` for basic functionality.
+ *
+ * `update` differs from `evaluate` in that `update` should be used when some
+ * LHS `Tensor` has been partially computed, and now we would like to update it
+ * with a RHS expression that contains it. In other words, this should be used
+ * when we would like to emulate assignment operations like `LHS +=`, `LHS -=`,
+ * `LHS *=`, etc.
+ *
+ * ### Example usage
+ * The equation \f$L_j = G R^{i}{}_{jl} S^{l}{}_{i} + H_j\f$ can be broken
+ * up into an initial call to `evaluate` to compute
+ * \f$R^{i}{}_{jl} S^{l}{}_{i}\f$, followed by a call to `update` to compute the
+ * rest of the expression:
+ * \code{.cpp}
+ * auto L = TensorExpressions::evaluate<ti_j>(
+ *     R(ti_I, ti_j, ti_l) * S(ti_L, ti_i));
+ *
+ * TensorExpressions::update<ti_j>(
+ *     make_not_null(&L), G() * L(ti_j) + H(ti_j));
+ * \endcode
+ *
+ * One important difference to note with `update` is that it cannot split up the
+ * RHS expression and evaluate subtrees, while `evaluate` can (see
+ * `TensorExpression`). From benchmarking, it was found that the runtime of
```

Comment:
- Could you add a few comments what the issue with subtrees here is?
- Also, I didn't realize that `evaluate` allows the result tensor as one of the operands in the TE. Is that always safe? E.g., what happens if you compute:

    ```cpp
    evaluate<ti_i, ti_j>(make_not_null(&A), A(ti_j, ti_i));
    ```

    To me it seems nontrivial to make sure permutations like this work correctly with only the result buffer available. Am I wrong, or is workspace memory allocated somewhere?
- Lastly, if update operations like the one above are indeed safe, I think it would be nice to just use `evaluate` for this instead of the extra `update` function (which only calls back into `evaluate` anyway). Is it possible to figure out the `EvaluateSubtrees` flag without the user's help?

If the answer is that these update operations are _not_ safe (or only a subset of them are safe), then I think it's no problem to just have the user compute temporaries themselves in separate variables (workspace memory). It would be good to assert which operations are safe and which are not though.

Reply:
- Yes, I'll add some comments! The short answer is that if the LHS tensor appears in the RHS expression, we can't use the splitting subtree stuff because that traversal updates the LHS tensor with each subtree that's evaluated (the "accumulation" that's mentioned in the `TensorExpression` class documentation), so a term in RHS would change during the traversal, which is yikes ❌ because the LHS won't be computed correctly.
- You make a good point with the example, which is a case I forgot to consider. If the indices used for the LHS tensor on both the LHS and the RHS are the same, it should be safe. I think I should add a `static_assert` that checks that the generic indices are the same.
- If the generic indices of the LHS tensor used on both sides are the same, then it's safe, but otherwise it's not.

It's possible to eliminate `update`, have `evaluate` just look and see if the LHS tensor is in the RHS expression and, if so, then don't do the split subtree traversal, but the issue is that I believe both tree traversals (the split traversal with `evaluate_primary_subtree`/`get_primary` and the one-liner traversal with `get`) will be compiled because whether or not the LHS tensor is in the RHS expression can only be known at runtime, as opposed to compile time. Compiling both versions isn't ideal because it adds bloat to compile time and in cases where we have a large TE that doesn't have the LHS tensor in the RHS, we definitely wouldn't want the one-liner traversal compiled because `DataVector` expression compile time scales poorly with the number of operations. By having `EvaluateSubtrees`, I think only one of the traversals will be compiled. This was the reasoning for defining a separate function, but let me know if you have follow up thoughts!

Reply:
- Yes, I'll add some comments! The short answer is that if the LHS tensor appears in the RHS expression, we can't use the splitting subtree stuff because that traversal updates the LHS tensor with each subtree that's evaluated (the "accumulation" that's mentioned in the `TensorExpression` class documentation), so a term in RHS would change during the traversal, which is yikes ❌ because the LHS won't be computed correctly.
- You make a good point with the example, which is a case I forgot to consider. If the indices used for the LHS tensor on both the LHS and the RHS are the same, it should be safe. I think I should add a `static_assert` that checks that the generic indices are the same.
- If the generic indices of the LHS tensor used on both sides are the same, then it's safe, but otherwise it's not.

It's possible to eliminate `update`, and instead have `evaluate` just look and see if the LHS tensor is in the RHS expression and, if so, then do the one-liner traversal instead of the split subtree traversal. But by having `update`/`EvaluateSubtrees` (something known at compile time), and we can `static_assert` that if people want to use the LHS tensor in the RHS expression, it will be handled safely. Moreover, to your point about the example you gave where the indices of the LHS tensor were different on the LHS and RHS, this is something we can `static_assert` not to do at compile time with `update/EvaluateSubtrees`, but only at runtime with `evaluate` if we check for the LHS int he RHS. This was the reasoning for defining a separate function, but let me know if you have follow up thoughts!

Reply:
Addressed by new fixups https://github.com/sxs-collaboration/spectre/pull/3842/commits/7adc3cd3837d1eac47e914b3a3fd6f967a53ae21 (asserting same index order for LHS used in RHS) and https://github.com/sxs-collaboration/spectre/pull/3842/commits/7adc3cd3837d1eac47e914b3a3fd6f967a53ae21 (documenting why it's safe)